mitiq/pec/representations/learning.py

```diff
@@ -0,0 +1,152 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+"""Function to calculate parameters for biased noise model via a
+learning-based technique."""
+
+from typing import Optional, List
+import numpy as np
+from scipy.optimize import minimize
+from cirq import Circuit, LineQubit
+from mitiq import QPROGRAM, Executor, Observable
+from mitiq.cdr import generate_training_circuits
+from mitiq.pec import execute_with_pec
+from mitiq.pec.representations.biased_noise import (
+    represent_operation_with_local_biased_noise,
+)
+
+
+def learn_biased_noise_parameters(
+    operation: QPROGRAM,
+    circuit: QPROGRAM,
+    ideal_executor: Executor,
+    noisy_executor: Executor,
+    num_training_circuits: int = 10,
+    epsilon0: float = 0,
+    eta0: float = 1,
+    observable: Optional[Observable] = None,
+) -> List[float]:
+    r"""Loss function: optimize the quasiprobability representation using
+    the method of least squares
+
+    Args:
+        operation: ideal operation to be represented by a (learning-optmized)
+            combination of noisy operations.
+        circuit: the full quantum program as defined by the user.
+        ideal_executor: Executes the ideal circuit and returns a
+            `QuantumResult`.
+        noisy_executor: Executes the noisy circuit and returns a
+            `QuantumResult`.
+        num_training_circuits: number of Clifford circuits to be generated for
+            training data.
+        epsilon0: initial guess for noise strength.
+        eta0: initial guess for noise bias.
+        observable (optional): Observable to compute the expectation value of.
+            If None, the `executor` must return an expectation value. Otherwise
+            the `QuantumResult` returned by `executor` is used to compute the
+            expectation of the observable.
+
+    Returns:
+        Optimized noise strength epsilon and noise bias eta.
+    """
+    training_circuits = generate_training_circuits(
+        circuit=circuit,
+        num_training_circuits=num_training_circuits,
+        fraction_non_clifford=0,
+        method_select="uniform",
+        method_replace="closest",
+    )
+
+    ideal_values = np.array(
+        [ideal_executor.evaluate(t, observable) for t in training_circuits]
+    )
+
+    x0 = np.array(
+        [epsilon0, eta0]
+    )  # initial parameter values for optimization
+    result = minimize(
+        biased_noise_loss_function,
+        x0,
+        args=(
+            operation,
+            circuit,
+            ideal_values,
+            noisy_executor,
+            num_training_circuits,
+            observable,
+        ),
+        method="BFGS",
+    )
+    x_result = result.x
+    epsilon = x_result[0]
+    eta = x_result[1]
+
+    return [epsilon, eta]
+
+
+def biased_noise_loss_function(
+    x0: np.ndarray,
+    operation: QPROGRAM,
+    circuit: QPROGRAM,
```

Comment:
This is more a conceptual question. Shouldn't `biased_noise_loss_function` take a list of `training_circuits`  as argument, such that we can get the associated list of `mitigated_values` to be compared with `ideal_values`?

At the moment we compare the `mitigated_value` of `circuit` which is not necessarily close to the ideal results of  `training_circuits`

Reply:
The list of `training_circuits` is used to calculate the list of `ideal_values` before the loop (so outside of the loss function). Then the loss function compares the list of `mitigated_values` with the list of `ideal_values`.

The confusion might have come from the `num_training_circuits`  argument of the loss function, which actually isn't needed. Now I'm using `len(ideal_values)` in its place.